File: placement-training/5th_Jan/queue.c
#include <stdio.h>
#define MAX 5

int queue[MAX];
int front = -1;
int rear  = -1;

int isEmpty() {
    return (front == -1);
}

int isFull() {
    return (rear == MAX - 1);
}

void enqueue(int item) {
    if (isFull()) {
        printf("Queue Overflow\n");
        return;
    }
    if (front == -1) {
        front = 0;
    }
    rear++;
    queue[rear] = item;
    printf("Inserted: %d\n", item);
}

int dequeue() {
    int item;
    if (isEmpty()) {
        printf("Queue Underflow\n");
        return -1;
    }
    item = queue[front];
    front++;

    if (front > rear) {
        front = -1;
        rear  = -1;
    }
    return item;
}

int peek() {
    if (isEmpty()) {
        printf("Queue is empty\n");
        return -1;
    }
    return queue[front];
}

void display() {
    int i;
    if (isEmpty()) {
        printf("Queue is empty\n");
        return;
    }
    printf("Queue elements: ");
    for (i = front; i <= rear; i++) {
        printf("%d ", queue[i]);
    }
    printf("\n");
}

int main() {
    enqueue(10);
    enqueue(20);
    enqueue(30);

    display();

    printf("Deleted: %d\n", dequeue());
    printf("Front element: %d\n", peek());

    display();
    return 0;
}

File: placement-training/queues/Queue.java
public class Queue {
    private int[] queue;
    private int front, rear, capacity;

    public Queue(int size) {
        capacity = size;
        queue = new int[capacity];
        front = -1;
        rear = -1;
    }

    public boolean isEmpty() { return front == -1; }
    public boolean isFull() { return rear == capacity - 1; }

    public void enqueue(int item) {
        if (isFull()) { System.out.println("Queue Overflow"); return; }
        if (front == -1) front = 0;
        queue[++rear] = item;
        System.out.println("Inserted: " + item);
    }

    public int dequeue() {
        if (isEmpty()) { System.out.println("Queue Underflow"); return -1; }
        int item = queue[front++];
        if (front > rear) { front = -1; rear = -1; }
        return item;
    }

    public int peek() {
        if (isEmpty()) { System.out.println("Queue is empty"); return -1; }
        return queue[front];
    }

    public void display() {
        if (isEmpty()) { System.out.println("Queue is empty"); return; }
        System.out.print("Queue: ");
        for (int i = front; i <= rear; i++) System.out.print(queue[i] + " ");
        System.out.println();
    }

    public static void main(String[] args) {
        Queue q = new Queue(5);
        q.enqueue(10);
        q.enqueue(20);
        q.enqueue(30);
        q.display();
        System.out.println("Deleted: " + q.dequeue());
        System.out.println("Front: " + q.peek());
        q.display();
    }
}
